uploaders/common.go

```diff
@@ -110,32 +110,60 @@ func createArtifact(buildURL, token, artifactPth, artifactType string) (string,
 func uploadArtifact(uploadURL, artifactPth, contentType string) error {
 	log.Printf("uploading artifact")
 
-	data, err := os.Open(artifactPth)
-	if err != nil {
-		return fmt.Errorf("failed to open artifact, error: %s", err)
+	netClient := &http.Client{
+		Timeout: 10 * time.Minute,
 	}
-	defer func() {
-		if err := data.Close(); err != nil {
-			log.Errorf("Failed to close file, error: %s", err)
+
+	return retry.Times(3).Wait(5).Try(func(attempt uint) error {
+		file, err := os.Open(artifactPth)
+		if err != nil {
+			return fmt.Errorf("failed to open artifact, error: %s", err)
 		}
-	}()
+		defer func() {
+			if err := file.Close(); err != nil {
+				log.Warnf("failed to close file, error: %s", err)
+			}
+		}()
 
-	args := []string{"curl", "--fail", "--tlsv1", "--globoff"}
-	if contentType != "" {
-		args = append(args, "-H", fmt.Sprintf("Content-Type: %s", contentType))
-	}
-	args = append(args, "-T", artifactPth, "-X", "PUT", uploadURL)
+		request, err := http.NewRequest(http.MethodPut, uploadURL, ioutil.NopCloser(file))
+		if err != nil {
+			return fmt.Errorf("failed to create request, error: %s", err)
+		}
 
-	return retry.Times(3).Wait(5 * time.Second).Try(func(attempt uint) error {
-		if attempt > 0 {
-			log.Warnf("%d attempt failed", attempt)
+		request.Header.Add("Content-Type", contentType)
+
+		// Set Content Length manually (https://stackoverflow.com/a/39764726), as it is part of signature in signed URL
+		fileInfo, err := file.Stat()
+		if err != nil {
+			return fmt.Errorf("failed to get file info for %s, error: %s", artifactPth, err)
 		}
-		cmd, err := command.NewFromSlice(args)
+		request.ContentLength = fileInfo.Size()
+
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
```

Comment:
The client has 10 mins of timeout, but the request could be called 3 times: if the first call timeouts it would not be called more, as the clients timeout already expired.
Why not set timeout to netClient.Timeout/3?

Reply:
the client's timeout is per request: https://golang.org/pkg/net/http/#Client

Reply:
Right, don't know why I messed it up. Note: it would be still better to use netClient.timeout, as  if you want to change the duration, you have to change it in one place, but this is not that important.